### PR TITLE
gitbase: implement sql.Indexable on all tables

### PR DESCRIPTION
**Pull request by erizocosmico**
Most of the contents in #279 did not survive because they relied on the index being something generic for all tables when in reality, almost every table has its own logic regarding the index storage.

This implementation differs *a lot* from what was initially thought in #295. The way the indexes work made it impossible to work that way.
There is also a lot of code that is inside go-git, but that logic is not exposed. If it's ever exposed we might want to refactor some of that.

To sum it up, the implementation looks like this:

- Every table has its own key value iterator with its own index key, because most of them need different things stored so they can be converted to rows efficiently after getting the key from the index.
- There is some common logic to find packfiles for a given repository and get from there the offsets of the objects that will be indexed.
- For `ref_commits`, `commit_trees` and `commit_blobs` there is a common logic `rowKeyValueIter` and `rowIndexIter`, since they are stored as is. The other tables had no obvious common logic, so instead of coming up with an abstraction that might be a PITA in the future, I preferred to have some more tedious and perhaps slightly repeating implementations on the other iterators (though there are no exact identical impls).

### CODE REVIEW

**Comment by ajnavarro** [comment]
Related: https://github.com/src-d/go-git/pull/854

File: commits.go [hunk 1]
```diff
@@ -261,3 +294,107 @@ func (i *commitsByHashIter) nextList() (*object.Commit, error) {
 		return commit, nil
 	}
 }
+
+type commitIndexKey struct {
```
Comment (ajnavarro) [inline commits.go]: Could we have a generic `ObjectIndexKey` struct for all the objects that are into the packfile?

**Comment by erizocosmico** [comment]
Only commit and blob have the struct totally in common. Tree entries also needs the pos inside the tree and the rest of the tables are very different. I can make one common object for blobs and commits tho.
On Tue, 5 Jun 2018 at 18:51, Antonio Navarro Perez <notifications@github.com<mailto:notifications@github.com>> wrote:

@ajnavarro commented on this pull request.

________________________________

In commits.go<https://github.com/src-d/gitbase/pull/298#discussion_r193145645>:

> @@ -261,3 +294,107 @@ func (i *commitsByHashIter) nextList() (*object.Commit, error) {
                return commit, nil
        }
 }
+
+type commitIndexKey struct {


Could we have a generic ObjectIndexKey struct for all the objects that are into the packfile?

**Comment by erizocosmico** [comment]
Added the tests and some fixes, along with the unification of the index keys where it was possible. It can be reviewed now.

**Comment by ajnavarro** [comment]
go-git 4.4.1 has dotgit logic outside `internal` folder, so we can use it now updating the dependency.

**Comment by erizocosmico** [comment]
@ajnavarro Since this is against a feature branch and won't go straight to master, what do you think if we review and merge this one first and then in a followup PR I update the dependency and replace the manual impl with the dotgit one?

File: packfiles.go [hunk 2]
```diff
@@ -199,15 +199,15 @@ func (i *objectIter) Next() (*encodedObject, error) {
...
 			RepositoryID: i.packs.repo.path,
-			Packfile:     i.packs.packfiles[i.packs.packpos],
+			Packfile:     i.packs.packfiles[i.packs.packpos-1],
 		}, nil
 	}
 }
 
 func (i *objectIter) Close() error {
-	if i.packObjects != nil {
+	/*if i.packObjects != nil {
```
[the remark this answers is not in the document]
Comment (erizocosmico) [reply packfiles.go]: yep

**Comment by erizocosmico** [comment]
Fixed errors in travis

File: cmd/gitbase/command/server.go [hunk 3]
```diff
@@ -28,13 +32,15 @@ const (
...
-	Password string   `short:"P" long:"password" default:"" description:"Password used for connection"`
+	Verbose   bool     `short:"v" description:"Activates the verbose mode"`
+	Git       []string `short:"g" long:"git" description:"Path where the git repositories are located, multiple directories can be defined. Accepts globs."`
+	Siva      []string `long:"siva" description:"Path where the siva repositories are located, multiple directories can be defined. Accepts globs."`
+	Host      string   `short:"h" long:"host" default:"localhost" description:"Host where the server is going to listen"`
+	Port      int      `short:"p" long:"port" default:"3306" description:"Port where the server is going to listen"`
+	User      string   `short:"u" long:"user" default:"root" description:"User name used for connection"`
+	Password  string   `short:"P" long:"password" default:"" description:"Password used for connection"`
+	PilosaURL string   `long:"pilosa" default:"http://localhost:10101" description:"URL to your pilosa server"`
+	IndexDir  string   `short:"i" long:"index" default:"/var/gitbase/index" description:"Directory where the gitbase indexes information will be persisted."`
```
Comment (ajnavarro) [inline cmd/gitbase/command/server.go]: Maybe we should change the default path. What do you think @jfontan ?

File: cmd/gitbase/command/server.go [hunk 4]
```diff
@@ -104,6 +110,18 @@ func (c *Server) buildDatabase() error {
 	c.engine.Catalog.RegisterFunctions(function.Functions)
 	logrus.Debug("registered all available functions in catalog")
 
+	if err := os.MkdirAll(c.IndexDir, 0755); err != nil {
```
Comment (ajnavarro) [inline cmd/gitbase/command/server.go]: Should we add some log on these new steps?

File: files.go [hunk 5]
```diff
@@ -251,3 +299,195 @@ func fileToRow(
...
+		values, err := rowIndexValues(row, i.columns, FilesSchema)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		return values, key, nil
+	}
+}
+
+func (i *filesKeyValueIter) Close() error {
```
Comment (ajnavarro) [inline files.go]: We should call this on tests too.

File: repositories.go [hunk 6]
```diff
@@ -93,6 +94,46 @@ func (r *repositoriesTable) WithProjectAndFilters(
...
+	span, ctx := ctx.Span("gitbase.RepositoriesTable.WithProjectFiltersAndIndex")
+	s, ok := ctx.Session.(*Session)
+	if !ok || s == nil {
+		span.Finish()
+		return nil, ErrInvalidGitbaseSession.New(ctx.Session)
+	}
+
+	var iter sql.RowIter = &rowIndexIter{index}
+
+	if len(filters) > 0 {
```
Comment (ajnavarro) [inline repositories.go]: Can we test this with some filter?

**Comment by ajnavarro** [comment]
Check codecov, appears that squash_join is not tested at all.

File: cmd/gitbase/command/server.go (hunk 3 above)
Comment (erizocosmico) [reply cmd/gitbase/command/server.go]: What path should we change it to?

Comment (jfontan) [reply cmd/gitbase/command/server.go]: Usually software does not have var directories directly in `/var` but in one of its subdirectories. MySQL has its database files in `/var/lib/mysql`.

**Comment by erizocosmico** [comment]
Fixed the coverage thing of the squash tables

File: remotes.go [hunk 7]
```diff
@@ -99,6 +101,46 @@ func (r *remotesTable) WithProjectAndFilters(
...
+	return &remotesKeyValueIter{repos: iter, columns: colNames}, nil
+}
+
+// WithProjectFiltersAndIndex implements sql.Indexable interface.
+func (*remotesTable) WithProjectFiltersAndIndex(
+	ctx *sql.Context,
+	columns, filters []sql.Expression,
+	index sql.IndexValueIter,
+) (sql.RowIter, error) {
+	span, ctx := ctx.Span("gitbase.ReferencesTable.WithProjectFiltersAndIndex")
```
Comment (jfontan) [inline remotes.go]: Shouldn't this be `"gitbase.RemotesTable.WithProjectFiltersAndIndex"`?

Comment (erizocosmico) [reply remotes.go]: Fixed

**Comment by erizocosmico** [comment]
Everything is fixed now, upgraded go-mysql-server and added engine initialization to the server command.

**Pull request merged**